sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/TypeInference.scala

```diff
@@ -0,0 +1,429 @@
+  def isJavaDate: Boolean = false
+  def isJavaInstant: Boolean = false
+  def isJavaTimestamp: Boolean = false
+  def isJavaString: Boolean = false
+  def isJavaBigDecimal: Boolean = false
+  def isJavaBigInteger: Boolean = false
+
+  def isScalaBigDecimal: Boolean = false
+  def isScalaBigInt: Boolean = false
+
+  /**
+   * Defines deserializers for specified `TypeInference`.
+   */
+  def deserializerFor(path: Expression, walkedTypePath: WalkedTypePath): Expression
+}
+
+case class ScalaInference(scalaType: `Type`) extends TypeInference {
+
+  val t = scalaType.dealias
+
+  override def isNotObjectType: Boolean = !dataTypeFor(t).isInstanceOf[ObjectType]
+
+  override def isJavaShort: Boolean = t <:< localTypeOf[java.lang.Short]
+  override def isJavaInteger: Boolean = t <:< localTypeOf[java.lang.Integer]
+  override def isJavaLong: Boolean = t <:< localTypeOf[java.lang.Long]
+  override def isJavaDouble: Boolean = t <:< localTypeOf[java.lang.Double]
+  override def isJavaFloat: Boolean = t <:< localTypeOf[java.lang.Float]
+  override def isJavaByte: Boolean = t <:< localTypeOf[java.lang.Byte]
+  override def isJavaBoolean: Boolean = t <:< localTypeOf[java.lang.Boolean]
+
+  override def isJavaLocalDate: Boolean = t <:< localTypeOf[java.time.LocalDate]
+  override def isJavaDate: Boolean = t <:< localTypeOf[java.sql.Date]
+  override def isJavaInstant: Boolean = t <:< localTypeOf[java.time.Instant]
+  override def isJavaTimestamp: Boolean = t <:< localTypeOf[java.sql.Timestamp]
+  override def isJavaString: Boolean = t <:< localTypeOf[java.lang.String]
+  override def isJavaBigDecimal: Boolean = t <:< localTypeOf[java.math.BigDecimal]
+  override def isJavaBigInteger: Boolean = t <:< localTypeOf[java.math.BigInteger]
+
+  override def isScalaBigDecimal: Boolean = t <:< localTypeOf[BigDecimal]
+  override def isScalaBigInt: Boolean = t <:< localTypeOf[scala.math.BigInt]
+
+  def isSQLUserDefinedType: Boolean =
+    t.typeSymbol.annotations.exists(_.tree.tpe =:= typeOf[SQLUserDefinedType])
+  def isRegisteredUDT: Boolean = UDTRegistration.exists(getClassNameFromType(t))
+
+  def isScalaArray: Boolean = t <:< localTypeOf[Array[_]]
+  def isScalaOption: Boolean = t <:< localTypeOf[Option[_]]
+  def isScalaSeq: Boolean = t <:< localTypeOf[Seq[_]]
+  def isScalaSet: Boolean = t <:< localTypeOf[scala.collection.Set[_]]
+  def isScalaMap: Boolean = t <:< localTypeOf[Map[_, _]]
+  def isDefinedByConstructorParams: Boolean = ScalaReflection.definedByConstructorParams(t)
+
+  override def deserializerFor(
+      path: Expression,
+      walkedTypePath: WalkedTypePath): Expression = this match {
+    case _ if isScalaOption =>
+      val TypeRef(_, _, Seq(optType)) = t
+      val className = getClassNameFromType(optType)
+      val newTypePath = walkedTypePath.recordOption(className)
+      WrapOption(TypeInference.deserializerFor(ScalaInference(optType), path, newTypePath),
+        dataTypeFor(optType))
+
+    case _ if isScalaArray =>
```

Comment:
I feel this can be consolidated further with Java's array and list cases. But not do it now in order to reduce diff.